### ISSUE TITLE
Wrong technical validity timestamp of EU digital certificates

### ISSUE THREAD

**Issue by apfarr** [corona-warn-app/cwa-documentation#689]
Hi all,

I believe there is a bug concerning the technical expiration timestamp for rapid antigen test certificates. Since I'm just a user I'm not sure on which side the bug is. Or maybe there are two bugs...

My family used the Ecocare provider for rapid antigen tests several times. When you book an appointment you can give a consent that the result should be pushed to CWA. That works fine: CWA prompts to add the certificate when the test result is there.

Sometimes adding a certificate fails saying that it is not a valid certificate. Sometimes a certificate is added successfully but when you validate it in CWA or another app (some countries like France or Latvia have designated apps for that) the certificate is not recognized as valid.

I looked into the issue and noticed that the technical validity period is kept extremely short. E.g. the test was taken at 11:03 and the technical validity expired at 11:44 (same day). In another case the test was taken at 15:29 and the technical validity expired at 19:00 (same day).

In fact, if I select a date and time within the technical validity period, the validation is successful.

I suppose that the timestamp of technical expiration is the moment of the attempt of adding the certificate to CWA. In the example above, the test was taken at 11:03; the result was certified at RKI and made ready for CWA at 11:20; and 11:44 was the time when I opened CWA and tried to add the certificate. In the example of 15:29/19:00 I'm even more sure: my wife reported by Whatsapp that she could not add her certificate on her phone ("QR code is not a valid EU digital certificate" or like that), so I did it for her on my phone by scanning her QR code and sent her that screenshot -- and it was exactly 19:00.

It looks like an attempt to add a certificate on another phone makes the certificate technically invalid? If so, it's a joke... We have to add and manage EU digital certificates of our daughter (under 16) on our parents' phones and it never worked. For us adults it worked partially.

---
related:
Internal Tracking ID: [EXPOSUREAPP-8122](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-8122)
Internal Tracking ID: [EXPOSUREAPP-8156](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-8156)

**Comment by vaubaehn** [comment]
I don't have much time, but I'm jumping in here quickly:
CWA does NOT create the certificate. The certificate is created by the test provider (here: EcoCare). Test provider calculates hash of certificate, sends it to TSI's backend, TSI sends it to IBM server for signing. Test provider pulls for the signature, until it's available. When the signature is received by the test provider, all data are combined (DCC + signature) and sent to CWA test result server (also maintained by TSI). CWA pulls the whole certificate, as soon it is available.

So:
>     1. Ecocare sends wrong expiration timestamps - the complaint in the linked issue also concerned this provider.

is right. And as far as I remember, it's not the first time, they have problems here...

Two parties are responsible:
EcoCare (or any other test provider) to correctly assemble the certificates.
TSI for validating the certificates during the "Abnahmetest", before the test provider is connected to the backend. When being in production state, DCCs from test provider can't be easily valdiated for technical accuracy anymore.

See also here for some technical details/links:
https://github.com/corona-warn-app/cwa-app-android/issues/3850#issuecomment-896904172

There are other unsolved issues in this regard:
https://github.com/corona-warn-app/cwa-documentation/issues/661
https://github.com/corona-warn-app/cwa-quicktest-onboarding/issues/24

Somehow it looks like, as TSI doesn't care too much to get this fixed and to improve Quality Assurance in this regard.
HTH

**Comment by vaubaehn** [comment]
> And as far as I remember, it's not the first time, they have problems here...

It was this issue: https://github.com/corona-warn-app/cwa-documentation/issues/668

@thomasaugsten Could you inform the test provider, please? Imho, that would be TSI's job, but they're probably too busy...

**Comment by dsarkar** [comment]
@apfarr Thanks again for the detailed OP. We definitely will have a look into this and forward this to the adequate entities. We will come back to you. Best, DS

---
Corona-Warn-App Open Source Team

**Comment by vaubaehn** [comment]
@dsarkar 🙏

**Comment by dsarkar** [comment]
@vaubaehn Thanks also for stepping in here.

Related:
Internal Tracking ID: [EXPOSUREAPP-8122](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-8122)
Internal Tracking ID: [EXPOSUREAPP-8156](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-8156)

TSI has been contacted.

**Comment by apfarr** [comment]
Thanks vaubaehn!

If you diagnose that "Ecocare sends wrong expiration timestamps" I wonder how can it happen that such a wrong timestamp always corresponds to the moment when the user tries to add the certificate in CWA? These are not random timestamps.

**Comment by Ein-Tim** [comment]
@vaubaehn 

Thanks for stepping in here and clearing things up! ❤️

**Comment by vaubaehn** [comment]
@apfarr 
You're welcome!

> If you diagnose that "Ecocare sends wrong expiration timestamps" I wonder how can it happen that such a wrong timestamp always corresponds to the moment when the user tries to add the certificate in CWA? These are not random timestamps.

They have a flaw in their scripts.
Please read here (and the following comments) first: https://github.com/corona-warn-app/cwa-documentation/issues/661#issuecomment-877199316
This handles the timestamp for the RAT test in CWA (independent from DCC). But this timestamp can be re-used to create/assemble the DCC from the test provider.

**Comment by apfarr** [comment]
And I'd still like to repeat a question to minimize the error probability on the user's side:

Is the process of adding a certificate bound to the device where the consent to submit data to RKI was given? Is it possible to give the consent (during appointment booking) on one device but add the certificate by scanning the QR code in the result on another device?

**Comment by apfarr** [comment]
@vaubaehn: Alright. Thx.

I also complained to Ecocare. I described it as in my initial posting here.

**Comment by vaubaehn** [comment]
> Is it possible to give the consent (during appointment booking) on one device but add the certificate by scanning the QR code in the result on another device?

The critical moment is, when the test (not the DCC) is registered from CWA. CWA sends a request to test result server and receives a unique token to pull the test later. This token can only be received once.
So, when you receive a link or QR code for a test during the booking, but you don't import it to CWA, then you could use another device to receive the result, when you forward the link/QR code to that device at a later time.

Because of the flaws with the 'sc' timestamp, that also interferes in the issue here, it _can_ be possible to manipulate the period of time that the (negative) result is valid.

**Comment by apfarr** [comment]
Very interesting, thanks. Means that for a better flexibility it's better not to give consent to CWA at the time of appointment booking at all because it will bind the retrieval to a particular device, but to scan the QR code from the result.

However I'm afraid it would not work with Ecocare as it seems that they produce QR codes only if the user has given his consent in CWA... no pre-registration in CWA - no QR code...

**Comment by vaubaehn** [comment]
> However I'm afraid it would not work with Ecocare as it seems that they produce QR codes only if the user has given his consent in CWA... no pre-registration in CWA - no QR code...

Yes, that should probably the case for all test providers now: Because without consent they don't get any refund for the free-of-charge citizen test (to prevent Abrechnungsbetrug).
This is since August 1st.
The problem with this confirmation timestamp is, that it is incorrectly sometimes used by some 3rd party providers as the timestamp when the test sample was collected (when someone turned the stick inside your nose). And because of some flaws implementation TSI <-> Test Provider, this confirmation timestamp is silently accepted.
In case of Ecocare they are now using it wrong as the expiry date... We don't know, if they also use it as the 'sample collected date'...

**Comment by vaubaehn** [comment]
One last thing:
> Means that for a better flexibility it's better not to give consent to CWA at the time of appointment booking

You must give consent to receive the result in CWA at the time of appointment booking. But if the booking system displays a link or a QR code, you would need to take a copy, and import it to CWA with a different device later.

**Comment by apfarr** [comment]
Alright, I'll keep you all updated if I hear something from Ecocare. 

Now I also realize why some tests are retrievable and some others not. Not only Ecocare results are affected by this, recently we were unable to retrieve the result from schnelltestonline.de . We book appointments for 2 people (adult + child) on the same device and give consent in CWA. I suppose that in the end only one test token is stored, i.e. the latest one, and the previous one gets lost. When the result is bound to a particular device but the token is lost, it cannot be retrieved at all, by no other device - and also not by the device where the consent was given...

Summary: Children must break the law (no CWA under 16) and give consent on their phones. I see no other option.

**Comment by apfarr** [comment]
> You must give consent to receive the result in CWA at the time of appointment booking. But if the booking system
> displays a link or a QR code, you would need to take a copy, and import it to CWA with a different device later.

Ok, but I've not seen this manual solution yet. Testing providers push it directly to CWA, no link/code is displayed.

Anyway, user MUST define on which device (s)he wants to retrieve the result at the time of booking, at least until October 11th when tests are not free anymore. And user has just one shot when retrieving the result. And user should not book more than one test per device.

**Comment by vaubaehn** [comment]
> Ok, but I've not seen this manual solution yet. Testing providers push it directly to CWA, no link/code is displayed.

In case of NETSYNO (technical infrastructure for 'dm'), they are displaying both: a link and a QR code, after you entered your data on their booking platform. You need to either click on the link _manually_ to import it into CWA, or scan the QR code.
If you copy the link, and paste it somewhere (and don't register it in CWA), you can register the test at a later time.
If this does not work on your device, then it could be, that the testprovider has an automatic refferer to the deeplink , and your device automatically opens it.

**Comment by vaubaehn** [comment]
By the way: if I understood you correctly, that already registering the test in CWA with EcoCare sometimes works, and sometimes not, then there is another issue:

Hi EcoCare - in case your're reading here.
Please, make sure that the URL for the deeplink is encoded in base64URL! and not in just base64, as base 64 creates chars like '/' which will break the import into CWA...
If you have more questions, feel free to open an issue in https://github.com/corona-warn-app/cwa-quicktest-onboarding/issues .
I guess, someone from the community can guide you there. Thanks!

**Comment by apfarr** [comment]
> the test in CWA with EcoCare sometimes works, and sometimes not

It is the case but I cannot identify the scheme. Except that it never works on an Android but works on iOS most of the times. 

However we were recently unable to add a certificate from another provider, too. On no device. Including the device where the consent was given.

**Comment by Ein-Tim** [comment]
Great "discussion" here, @apfarr I wanted to apologize that I spread "misinformation" earlier, sorry for any confusion caused!

**Comment by apfarr** [comment]
Another quite interesting observations: 
1) The technical validitiy expiration date is displayed in the iOS app but not in the Android App.
2) For a validity check, it is possible to set a date in the past in the iOS app but not in the Android App.

So basically it is possible to verify the source of the trouble on iOS only ;)

**Comment by Ein-Tim** [comment]
@apfarr Are you sure the Android device is on the latest version (2.7.1)?

**Comment by apfarr** [comment]
It was 2.6.1. In 2.7.1 the technical validitiy expiration date is displayed - but it is still not possible to set a past date/time to check the validity (that is to check if the certificate was valid at least for some time in the past).

**Comment by apfarr** [comment]
Still another question. If a test certificate is supposed to be added to one device only, how was I able to add same test result (from Ecocare, consent for a particular device was given) on 3 further devices by scanning the QR code from the initial device?

**Comment by Ein-Tim** [comment]
@apfarr 

You have to differentiate here. 

1. The test result - you can only add it to one device
2. The test certificates - you can scan the QR code from it on as many devices as you want

**Comment by apfarr** [comment]
Pardon, I don't quite understand. A test result is delivered in form of a test certificate, isn't it? If not, what is the (technical) difference?

**Comment by Ein-Tim** [comment]
@apfarr 

The EU DCC and the test result delivery, these are two different features. Since the first version of CWA, the user was able to register their PCR test with a QR code in the app and receive the test result (on the main screen). 
With version 2.4 the feature to receive test result certificates was added. But still, you scan the QR code on the main screen, receive your test result on the main screen and **additionally** you get your test certificates in the certificates tab. 

Did this help? 

Maybe these screenshots make it a bit more clear: https://www.coronawarn.app/en/screenshots/#ios_results
And this blog post: https://www.coronawarn.app/en/blog/2021-06-24-cwa-version-2-4/

**Comment by apfarr** [comment]
OMG. So, if the PDF test result file contains a QR code, is this code supposed to be scanned using the Start Screen tab or the Certificates tab?

**Comment by Ein-Tim** [comment]
@apfarr 

No idea, just try it (-;

Maybe it's also for something completely different 🤷🏻‍♂️

**Comment by apfarr** [comment]
I have tried it already. I have a PDF test result file (paid RAT - not Ecocare) containing a QR code for CWA which looks absolutely same as a certificate QR code but not accepted by any phone including the phone where the appointment was booked. Neither by the Start Screen scanner nor by Certificates scanner. So I wonder if QR codes which are printed on some (only paid?) test results are different from certificate QR codes.

Generally, I think I realize now how the error I'm faced with can happen (although I used wrong terminology in my inital post as I mixed up test results and certificates). A test result is a set of metadata pushed from the test provider via RKI to the device where the consent in CWA was given (another option, which is however not available for free RATs, is to receive the result to the device by scanning the QR code from the PDF result document). It is always the user who initiates the generation of a DCC on his device, and this is the moment when metadata are sent to RKI to generate/sign a DCC certificate, and this is the moment when current timestamp is submitted as the technical expiration date. In this case the certificate is generated/signed and received back - but expires immediately.

**Comment by Ein-Tim** [comment]
> I have tried it already. I have a PDF test result file (paid RAT - not Ecocare) containing a QR code for CWA which looks absolutely same as a certificate QR code but not accepted by any phone including the phone where the appointment was booked. Neither by the Start Screen scanner nor by Certificates scanner. So I wonder if QR codes which are printed on some (only paid?) test results are different from certificate QR codes.

Which page opens when you scan it with a normal QR code scanner?

**Comment by apfarr** [comment]
CWA / add test result opens. Alright, I now understand how to check it.

**Comment by Ein-Tim** [comment]
@apfarr You said that you can't add it to CWA, which error comes up?

**Comment by apfarr** [comment]
The error says that the QR code is invalid or the test is already registered on another smartphone.

However it is definitely not registered on another smartphone, all family members checked it and tried to add it on their phones. And as a user I can't check what is inside, whether it's invalid or not.

**Comment by Ein-Tim** [comment]
@apfarr 

Okay, so you did **not** receive the test result for which the PDF is via the Corona-Warn-App?

**Comment by apfarr** [comment]
No I didn't. 

Frankly speaking, I don't remember if we were prompted to give the consent during booking or not... it was a paid RAT.

**Comment by Ein-Tim** [comment]
@apfarr Okay, at this point, I suggest to get in touch with the RAT provider, since this could be seen as a privacy breach. Someone might have gotten unauthorized access to your test result, because obviously, the CWA code says that it was already used.

Maybe @thomasaugsten could give some advice what to do in such a case?

**Comment by vaubaehn** [comment]
@Ein-Tim 
That error message is shown, when the QR code is invalid (or has already been registered before).
The most likely cases:
1. It's not a QR code for registering the test in CWA, or
2. the test provider has a flaw in the creation of the QR code.Test provider must use base64url to encode/create the deeplink. If base64 is used, the link/encoded payload may randomly be invalid. One hint we already have that this could be the case, is, that this QR then usually works with Iphone (do you remember the issue with the broken event QR from a swimming hall some weeks ago? Similar problem).

**Comment by apfarr** [comment]
Ok, I'll try to sort it with the provider. We have both iPhones and Androids, it failed everywhere.

**Comment by apfarr** [comment]
Meanwhile the following question: the birth date is not validated anywhere, is it? Because it was the test result of a minor (who is not supposed to use CWA).

**Comment by Ein-Tim** [comment]
@vaubaehn 

> It's not a QR code for registering the test in CWA

This seems impossible since @apfarr said that scanning the QR code with a normal QR code scanner leads to the Corona Warn App (probably the URL started with s.coronawarn.app, or @apfarr?).

> the test provider has a flaw in the creation of the QR code.Test provider must use base64url to encode/create the deeplink. If base64 is used, the link/encoded payload may randomly be invalid. One hint we already have that this could be the case, is, that this QR then usually works with Iphone (do you remember the issue with the broken event QR from a swimming hall some weeks ago? Similar problem).

If this would be the case, it's so bad implemented that it fails under iOS & Android now (see https://github.com/corona-warn-app/cwa-documentation/issues/689#issuecomment-903160378)

> Meanwhile the following question: the birth date is not validated anywhere, is it? Because it was the test result of a minor (who is not supposed to use CWA).

No, the date of birth is not validated.

**Comment by apfarr** [comment]
> probably the URL started with s.coronawarn.app, or @apfarr?

Yes. If I use iOS embedded QR code scanner it does not display the link but prompts to open CWA. If I use a 3rd party scanner app, it displays a link starting with https://s.coronawarn.app/?v=1#<verylongkey> . I don't see any corrupted characters.

**Comment by Ein-Tim** [comment]
> Yes. If I use iOS embedded QR code scanner it does not display the link but prompts to open CWA. If I use a 3rd party scanner app, it displays a link starting with s.coronawarn.app/?v=1# . I don't see any corrupted characters.

Yep, then this is definitely a QR code intended to be used in the Corona-Warn-App.

**Comment by apfarr** [comment]
I have a guess. What happens if CWA has a stored token awaiting a test result but during this waiting a QR code from another test result is scanned?

**Comment by apfarr** [comment]
> then this is definitely a QR code intended to be used in the Corona-Warn-App.

Alright, now there is the answer: the QR codes of results and certificates look absolutely same but  are different.

The QR code of the result is just a hyperlink to s.coronawarn.app

The QR code of the certificate is not a link. It starts with the HC1: prefix followed by encrypted values divided by asterisks.

**Comment by Ein-Tim** [comment]
> I have a guess. What happens if CWA has a stored token awaiting a test result but during this waiting a QR code from another test result is scanned?

@apfarr While you have a test result registered and it's pending, you will see the "Your result is not yet available" box on the Home Screen. When you attempt to add a new test you will be informed that the old one will get deleted. Here's a screen recording:

https://user-images.githubusercontent.com/67682506/118037980-d4a09400-b36e-11eb-831b-42d11a1090ad.mp4

It's not relevant whether this is a PCR (as in the screen recording) or a rapid antigen test.

**Comment by apfarr** [comment]
Addendum: the QR code Ecocare attached to a result is not a result code but a certificate code. That's why it appears only after the CWA user generated the certificate. It means that Ecocare does not provide a QR code with the result link at all and the only option is to pre-register it via consent in CWA during booking.

**Comment by vaubaehn** [comment]
> If I use a 3rd party scanner app, it displays a link starting with https://s.coronawarn.app/?v=1#

After the '#' there should follow quite a lot more characters. When it has a wrong encoding (base64 instead of base64url) then sometimes '/' characters are encoded. This will break the deeplink and corrupts the payload. (base64url encodes '_' instead '/').

> If this would be the case, it's so bad implemented that it fails under iOS & Android now (see [#689 (comment)](https://github.com/corona-warn-app/cwa-documentation/issues/689#issuecomment-903160378))

From the swimming hall issue I remembered, that IPhone scanner could read the wrong encoded deeplink, while Android failed. Maybe I mixed something up, or the implementation here in this issue has different flaws/is otherwise broken.

**Comment by apfarr** [comment]
> After the '#' there should follow quite a lot more characters. When it has a wrong encoding (base64 instead of base64url) 
> then sometimes '/' characters are encoded. This will break the deeplink and corrupts the payload. (base64url encodes '_' instead '/').

Yes there is a long character string but it does not contain something that looks like broken encoding (I'm proficient in recognizing it).

Another guess. How long is a test result available on the RKI server for retrieval? Because the test was carried out a week ago. 

Regarding the certificate generation, I have a general question. As vaubaehn wrote that there is a glitch in a script of the testing provider, I guess that a test result contains not only metadata but also a script responsible for generation of DCC. Wouldn't it be more error-prone to make this script generic and integrate it in CWA? The script would require test results in a pre-defined data structure. Instead of writing scipts, testing providers would just provide a conversion/mapping their data format to this data structure which is easier to implement and debug.

**Comment by vaubaehn** [comment]
@apfarr 
> Yes there is a long character string but it does not contain something that looks like broken encoding (I'm proficient in recognizing it).

If your that proficient to recognize broken characters in encoded data, then you're probably similar proficient to decode that data? It would be helpful, if you could provide the clear text of the data here. For fast decoding I recommend cryptii.com.

> How long is a test result available on the RKI server for retrieval?

21 days.

> I guess that a test result contains not only metadata but also a script responsible for generation of DCC

No script is included in the data of the test result.

> Wouldn't it be more error-prone to make this script generic

A script to generate a EU Digital Covid Certificate needs to be as specific to create such a certificate. It can't be 'general'.

> and integrate it in CWA?

I don't know how is your understanding of coding applications. The data in the EU DCC are different for every person/every event. This means, the data for the DCC need to be dynamically assembled within a pre-defined data structure. The scripts/applications of the test provider are doing this job: they insert data for/from one person into that data structure, for example the timestamps or the result ('positive/negative'). It is not possible that there is no application running at provider's site, otherwise no data could ever be sent to CWA.

> Instead of writing scipts, testing providers would just provide a conversion/mapping their data format to this data structure which is easier to implement and debug.

This conversion/mapping (to the pre-defined data structure) is done with scripts/applications, like I wrote above.
If you want to see it like this: EcoCare failed to convert/map their data correctly.

Did you already read any documentation that I provided you in the links above? I recommend CWA onboarding, and anything you can find in the CWA documentation repository (there is a link to the CWA test result server also).

**Comment by Ein-Tim** [comment]
I suggest to close this issue now. This is a more general problem and it is covered in multiple other issues.

**Comment by larswmh** [comment]
Closing as suggested.
